app/models/manager_refresh/save_collection/helper.rb

```diff
@@ -45,14 +46,22 @@ def save_inventory_collection!(inventory_collection, association)
       ActiveRecord::Base.transaction do
         association.find_each do |record|
           index = inventory_collection.object_index_with_keys(unique_index_keys, record)
-          if unique_db_indexes.include?(index) # Include on Set is O(1)
+          if unique_db_primary_keys.include?(record.id) # Include on Set is O(1)
+            # When we have a complex :through relations, it can easily happen a same result is returned multiple times.
+            # In this case we log a warning and ask for relation to return a correct distinct result. We don't want to
+            # do SELECT DISTINCT by default, since it's slow.
+            _log.warn("An :association of #{inventory_collection} is not returning a DISTINCT result, please make sure"\
+                      "the SQL query returns DISTINCT results.")
```

Comment:
Also the exception would be too much, since this can be data related. So it;s possible this will pass for specs, where we just don't test enough of the data combination and fail in real env, where data are more complex.

So in this case, we can make this work and write a nice warning that someone can notice. But we don't want to turn this to a blocker bug. :-)